### PR TITLE
Martin/ios 278 time machine layout is wrong for

### PR DESCRIPTION
**Pull request by MartinJibber**
Fixes and performance improvements to the time machine layout.

Scrolling is super smooth now.

https://user-images.githubusercontent.com/88675954/143147938-d4968400-85ae-420c-8669-cf2a909f03cd.MP4

### CODE REVIEW

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 1]
```diff
@@ -14,13 +14,25 @@ protocol TimeMachineCollectionViewLayoutDataSource: AnyObject {
     func getMessage(at indexPath: IndexPath) -> Messageable?
 }
 
+class TimeMachineCollectionViewLayoutInvalidationContext: UICollectionViewLayoutInvalidationContext {
+    /// If true, the z ranges for all the items should be recalculated.
+    var shouldRecalculateZRanges = true
+}
+
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:21]: Running all the sorting math every time the offset changed was not performant with large sets of messages.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 2]
```diff
@@ -80,17 +89,39 @@ class TimeMachineCollectionViewLayout: UICollectionViewLayout {
...
+        guard let customInvalidationContext = invalidationContext as? TimeMachineCollectionViewLayoutInvalidationContext else {
+            return invalidationContext
+        }
+
+        // Changing the bounds doesn't affect item z ranges.
+        customInvalidationContext.shouldRecalculateZRanges = false
+
+        return customInvalidationContext
+    }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:107]: If the layout is invalidated due to bounds change, there's no need to re-run the math. 

In contrast, if items are being added/deleted/moved, then all the math should be updated.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 3]
```diff
@@ -80,17 +89,39 @@ class TimeMachineCollectionViewLayout: UICollectionViewLayout {
...
+
+        guard let customContext = context as? TimeMachineCollectionViewLayoutInvalidationContext else { return }
 
         // Clear the layout attributes caches.
         self.cellLayoutAttributes.removeAll()
-        self.zRangesDict.removeAll()
         self.decorationLayoutAttributes.removeAll()
+        if customContext.shouldRecalculateZRanges {
+            self.zRangesDict.removeAll()
+        }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:119]: Now, when the layout is invalidated, we check to see if we should delete the z ranges cache.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 4]
```diff
@@ -80,17 +89,39 @@ class TimeMachineCollectionViewLayout: UICollectionViewLayout {
...
     }
 
     override func prepare() {
-        self.prepareZRanges()
+        // Don't recalculate z ranges if we already have them cached.
+        if self.zRangesDict.isEmpty {
+            self.prepareZRanges()
+        }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:126]: With this optimization, the time machine runs buttery smooth.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 5]
```diff
@@ -176,6 +209,9 @@ class TimeMachineCollectionViewLayout: UICollectionViewLayout {
         // All items in a section are positioned relative to its frontmost item.
         guard let frontmostIndexPath = self.getFrontmostIndexPath(in: indexPath.section) else { return nil }
 
+        // OPTIMIZATION: Don't calculate attributes for items that won't be visible.
+        guard abs(frontmostIndexPath.item - indexPath.item) < 4 else { return nil }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:215]: Another optimization to make it even faster.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 6]
```diff
@@ -126,7 +157,7 @@ class TimeMachineCollectionViewLayout: UICollectionViewLayout {
             let currentSection = indexPath.section
             let currentItemIndex = indexPath.item
 
-            var startZ: CGFloat = 0
+            var startZ: CGFloat = CGFloat(sortedItemsIndex) * self.itemHeight
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:162]: Fixes an issue where the second to last oldest message might look smaller than usual.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 7]
```diff
@@ -318,11 +359,15 @@ class TimeMachineCollectionViewLayout: UICollectionViewLayout {
         var frontmostIndexes: [IndexPath] = []
         for i in 0..<sectionCount {
             guard let frontmostIndex = self.getFrontmostIndexPath(in: i) else { continue }
-            frontmostIndexes.append(frontmostIndex)
+            guard let range = self.zRangesDict[frontmostIndex] else { continue }
+
+            if range.vector(to: self.zPosition) > -self.itemHeight {
+                frontmostIndexes.append(frontmostIndex)
+            }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:368]: Fixes an issue where the newest message wasn't highlighted for certain edge cases.

**Review by Benjamindodgson — APPROVED** [review]
Yeah... Im glad you are doing this.

**Pull request merged**